adapters/lora/src/main/java/org/eclipse/hono/adapter/lora/LoraCommandSubscriptions.java

```diff
@@ -0,0 +1,178 @@
+                        closeConsumersForGateway(notification.getTenantId(), notification.getDeviceId());
+                    }
+                });
+    }
+
+    @SuppressWarnings("rawtypes")
+    private void closeConsumersForGateway(final String tenantId, final String gatewayId) {
+        if (!commandSubscriptions.entrySet().stream().anyMatch(e -> e.getKey().getDeviceId().equals(gatewayId))) {
+            return;
+        }
+
+        final Span span = TracingHelper
+                .buildSpan(tracer, null, "gateway deleted, corresponding command consumer will be closed\"",
+                        getClass().getSimpleName())
+                .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_CLIENT)
+                .start();
+        TracingHelper.setDeviceTags(span, tenantId, gatewayId);
+        final SubscriptionKey key = new SubscriptionKey(tenantId, gatewayId);
+        final TriTuple<CommandConsumer, LoraProvider, Context> subscription = commandSubscriptions.get(key);
+        final List<Future> consumerCloseFutures = Collections.synchronizedList(new ArrayList<>());
+        final Context ctx = subscription.three();
+        ctx.runOnContext(x -> {
+            final CommandConsumer commandConsumer = subscription.one();
+            consumerCloseFutures.add(commandConsumer.close(null));
+        });
+        CompositeFuture.join(consumerCloseFutures)
+                .onComplete(x -> commandSubscriptions.remove(key))
+                .onComplete(x -> span.finish());
+    }
+
+    @SuppressWarnings("rawtypes")
+    private void closeConsumersForTenant(final String tenantId) {
+        if (!commandSubscriptions.entrySet().stream().anyMatch(e -> e.getKey().getTenant().equals(tenantId))) {
+            return;
+        }
+
+        final Span span = TracingHelper
+                .buildSpan(tracer, null, "check command subscriptions on tenant disabled/deleted event",
+                        getClass().getSimpleName())
+                .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_CLIENT)
+                .start();
+        TracingHelper.setDeviceTags(span, tenantId, null);
+        span.log("tenant disabled/deleted, corresponding command consumers will be closed");
+        final List<Future> consumerCloseFutures = Collections.synchronizedList(new ArrayList<>());
+        final List<SubscriptionKey> keysToRemove = Collections.synchronizedList(new ArrayList<>());
+        commandSubscriptions.entrySet().stream().filter(e -> e.getKey().getTenant().equals(tenantId)).forEach(e -> {
+            final Context ctx = e.getValue().three();
+            ctx.runOnContext(x -> {
+                final CommandConsumer commandConsumer = e.getValue().one();
+                consumerCloseFutures.add(commandConsumer.close(null));
+            });
+            keysToRemove.add(e.getKey());
+        });
+        CompositeFuture.join(consumerCloseFutures)
+                .onComplete(x -> keysToRemove.stream().forEach(commandSubscriptions::remove))
```

Comment:
(same as in closeConsumersForGateway)
`commandSubscriptions` entry can be removed in the iteration above already (using iterator), allowing the `keysToRemove` list to be removed.